src/Feature/Configuration/Feature_Configuration.re

```diff
@@ -14,49 +39,66 @@ let merge = model => {
     Config.Settings.union(Config.Schema.defaults(model.schema), model.user),
 };
 
-let initial = contributions =>
+let initial = (~configFile, contributions) =>
   merge({
     schema:
       Config.Schema.unionMany(
         contributions |> List.map(Config.Schema.fromList),
       ),
-    user: Config.Settings.empty,
+    user:
+      Internal.loadConfiguration(configFile)
+      |> Result.value(~default=Config.Settings.empty),
     merged: Config.Settings.empty,
   });
```

Comment:
This could have been implemented using init effects as well, and we'd usually want side-effects on init to be asynchronous anyway, but in this case I think blocking is the right approach. We don't want the "flash of unstyled content"-experience that might occur if we render before having loaded the configuration.

I also think we might want to pass in some kind of "configuration provider" instead of the configuration file directly, to give the consumer more control over loading, fallback etc. and better separation for testing. I'm just not sure how that would interact with the configuration commands, `ConfigurationFileChanged` event and such. Maybe we should have a separate `Feature_ConfigurationFile` project?

Reply:
Thanks for thinking about this!

> We don't want the "flash of unstyled content"-experience that might occur if we render before having loaded the configuration.

Definitely not.

> I also think we might want to pass in some kind of "configuration provider" instead of the configuration file directly, to give the consumer more control over loading, fallback etc. and better separation for testing

This would be great. We have a hacky-configuration-file-override in our integration tests: https://github.com/onivim/oni2/blob/9e84aedee6755919a95f11cdf773d4cb5e45b610/src/Store/ConfigurationStoreConnector.re#L14

It would be nice to generalize the 'provider' strategy and expose it at a higher-level.

>  I'm just not sure how that would interact with the configuration commands, ConfigurationFileChanged event and such. Maybe we should have a separate Feature_ConfigurationFile project?

I'm also curious how it would relate to the other configuration files we load:
- `keybindings.json`
- (soon) `persisted.json`

Would it make sense to have `Feature_KeybindingsFile` and `Feature_PersistedFile` - or should they be grouped / share common logic?

Reply:
I abstracted the config file out behind a `getUserSettings` function for now, just to make easy to manipulate the settings for the integration test. But in the future perhaps a `Service_ConfigurationFIle` project that allows subscribing to a configuration file would be a good approach. Might try that out to get a feel for using `luv` for file watching as well.